File: Solutions/Evaluate_Division/Evaluate_Division.py
#Second Fastest
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        
        if len(equations) == 0:
            return [-1.0 for i in range(len(queries))]
        if len(queries) == 0:
            return []
        
        self.dict1 = collections.defaultdict(dict)
        
        for i,equation in enumerate(equations):
            self.dict1[equation[0]][equation[1]] = values[i]
            self.dict1[equation[1]][equation[0]] = 1/values[i]
        self.result = []
        for query in queries:
            self.visited = set()
            temp = self.bfs(query[0],query[1],1)
            self.result.append(temp)
        
        return self.result
    
    def bfs(self,start,end,val):
        q = collections.deque()
        q.append((start,val))
        self.visited.add(start)
        if end not in self.dict1:
            return -1.0
        while q:
            s,val = q.popleft()
            if s not in self.dict1:
                return -1.0
            if s == end:
                return val  
            for news in self.dict1[s]:
                if news not in self.visited:
                    self.visited.add(news)
                    q.append((news,val*self.dict1[s][news]))
        return -1.0

#Fastest
from collections import defaultdict
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        result = [-1.0]*len(queries)
        # construct the graph
        graph = defaultdict(dict)
        for ind, (i, j) in enumerate(equations):
            graph[i][j]= values[ind]
            graph[j][i]= 1/values[ind]
            
        def helper(x, y):
            # find path from x to y, given a graph
            queue = [(x,1)]
            visited = set()
            while queue:
                node, val = queue.pop()
                if node in visited:
                    continue
                visited.add(node)
                neighbors = graph[node] # a dict {node:value}
                if y in neighbors:
                    return val*neighbors[y]
                else:
                    for n in neighbors:
                        queue.append((n, val*neighbors[n]))
                # queue = next_queue
            return -1.0
        
        for ind, (x, y) in enumerate(queries):
            if x not in graph or y not in graph:
                continue
            elif x==y and x in graph[x]:
                result[ind] = 1.0
            elif y in graph[x]:
                result[ind] = graph[x][y]
            else:
                # find the transition path
                # there maybe multiple paths, but since the assumption is there is no contradiction, it is safe to pick random one
                # depth first search, search neighbors and neighbors' neighbors until get the target
                
                result[ind] = helper(x, y)
        return result

File: Solutions/Evaluate_Division/Evaluate_Division.cpp
class Solution
{
private:
    unordered_map<string, vector<pair<string, double>>> graph; // adjacency list

    pair<bool, double> search(string &a, string &b, unordered_set<string> &vis, double val)
    {
        if (vis.count(a) == 0)
        {
            vis.insert(a); // mark a as visited
            for (auto p : graph[a])
            {
                double temp = val * p.second;
                if (p.first == b)
                {
                    return make_pair(true, temp);
                } // found result
                auto result = search(p.first, b, vis, temp);
                if (result.first)
                {
                    return result;
                } // found result
            }
        }
        return make_pair(false, -1.0);
    }

public:
    vector<double> calcEquation(vector<vector<string>> &eq, vector<double> &val, vector<vector<string>> q)
    {
        vector<double> ans;
        for (int i = 0; i < eq.size(); ++i)
        {
            graph[eq[i][0]].push_back(make_pair(eq[i][1], val[i]));       // build graph list a->b
            graph[eq[i][1]].push_back(make_pair(eq[i][0], 1.0 / val[i])); // build graph list b->a
        }
        for (auto p : q)
        {
            unordered_set<string> vis; // to record visited nodes
            if (p[0] == p[1] && graph.count(p[0]))
                ans.push_back(1.0);
            else
                ans.push_back(search(p[0], p[1], vis, 1.0).second);
        }
        return ans;
    }
};

//Fastest
class Solution
{
public:
    vector<double> calcEquation(vector<vector<string>> &equations, vector<double> &values, vector<vector<string>> &queries)
    {
        if (equations.empty() || queries.empty())
            return {};
        const int n = values.size();
        vector<double> res;
        unordered_map<string, unordered_map<string, double>> m;
        for (int i = 0; i < n; i++)
        {
            m[equations[i][0]].emplace(equations[i][1], values[i]);
            m[equations[i][1]].emplace(equations[i][0], 1 / values[i]);
            m[equations[i][0]].emplace(equations[i][0], 1.0);
            m[equations[i][0]].emplace(equations[i][1], 1.0);
        }
        for (auto query : queries)
        {
            if (!m.count(query[0]) || !m.count(query[1]))
            {
                res.push_back(-1.0);
                continue;
            }
            queue<pair<string, double>> q;
            unordered_set<string> visited{query[0]};
            q.push({query[0], 1.0});
            bool found = false;
            while (!q.empty() && !found)
            {
                auto pair = q.front();
                q.pop();
                //std::cout<<pair.first<<" "<<endl;
                if (query[1] == pair.first)
                {
                    found = true;
                    res.push_back(pair.second);
                    break;
                }
                for (auto t : m[pair.first])
                {
                    if (visited.count(t.first))
                        continue;
                    visited.insert(t.first);
                    t.second *= pair.second;
                    q.push(t);
                }
            }
            if (!found)
                res.push_back(-1.0);
        }
        return res;
    }
};